intmodule exporting (#3148)
chipsalliance/chisel: docs/src/explanations/intrinsics.md

---
layout: docs
title:  "Intrinsics"
section: "chisel3"
---

# Intrinsics

Chisel *Intrinsics* are used to instantiate implementation defined functionality. 
Intrinsics provide a way for specific compilers to extend the capabilities of
the language in ways which are not implementable with library code.

> Modules defined as a `IntModule` will be instantiated as normal modules, but the

Reviewer: ```suggestion
Modules defined as an `IntrinsicModule` will be instantiated as normal modules, but the
```

intrinsic field communicates to the compile what functionality to use to 
implement the module.  Implementations may not be as modules, the module nature
of intrinsics

Intrinsics will be typechecked by the implementation.  What intrinsics are 
available is documented by an implementation.

### Parameterization

Parameters can be passed as an argument to the IntModule constructor.

> 

Reviewer: can we show an example?

---
layout: docs

> title:  "Intrinsics"

Reviewer: There are 2 "table of contents" to add this -- the main explanations.md, and the menu.yml for the website. Can be done later if you don't mean to advertise this yet

section: "chisel3"
---

# Intrinsics

Chisel *Intrinsics* are used to instantiate implementation defined functionality. 
Intrinsics provide a way for specific compilers to extend the capabilities of
the language in ways which are not implementable with library code.

Modules defined as an `IntrinsicModule` will be instantiated as normal modules, 

> but the intrinsic field communicates to the compile what functionality to use to 

Reviewer 2: ```suggestion
but the intrinsic field communicates to the compiler what functionality to use to 
```

implement the module.  Implementations may not be actual modules, the module 
nature of intrinsics is merely for instantiation purposes.

Intrinsics will be typechecked by the implementation.  What intrinsics are 
available is documented by an implementation.

### Parameterization

Parameters can be passed as an argument to the IntModule constructor.

### Example

This following creates an intrinsic module for the intrinsic named 

> "TestIntrinsic".  It takes a parameter name "STRING" and has several ports.

Reviewer: "OtherIntrinsic"?

> "OtherIntrinsic".  It takes a parameter name "STRING" and has several ports.

Reviewer 2: I think this meant "named?"
```suggestion
"OtherIntrinsic".  It takes a parameter named "STRING" and has several ports.
```
